reactor-aeron-core/src/main/java/reactor/aeron/client/AeronClientConnector.java

```diff
@@ -1,406 +1,145 @@
-  }
-
-  /**
-   * Handler for complete signal from server. At the moment of writing this javadoc the server
-   * doesn't emit complete signal. Method is left with logging.
-   *
-   * @param sessionId session id
-   */
-  @Override
-  public void onDisconnect(long sessionId) {
-    logger.debug("Received DISCONNECT for sessionId: {}", category, sessionId);
-    dispose(sessionId);
-  }
-
-  @Override
-  public void onConnect(
-      long connectRequestId,
-      String clientChannel,
-      int clientControlStreamId,
-      int clientSessionStreamId) {
-    logger.error(
-        "CONNECT request is not supported, clientChannel: {}, "
-            + "clientControlStreamId: {}, clientSessionStreamId: {}",
-        clientChannel,
-        clientControlStreamId,
-        clientSessionStreamId);
-  }
-
-  /**
-   * ConnectAck promise. Holds pair of connect request UUID and associated MonoProcessor. Clients
-   * use method {@link #promise()} to subscribe. When server responds then method {@link
-   * #success(long, int)} is being called.
-   */
-  private class ConnectAckPromise implements Disposable {
-
-    private final long connectRequestId;
-    private final MonoProcessor<ConnectAckResponse> promise;
-
-    private ConnectAckPromise(long connectRequestId) {
-      this.connectRequestId = connectRequestId;
-      this.promise = MonoProcessor.create();
-    }
-
-    private Mono<ConnectAckResponse> promise() {
-      return promise;
-    }
-
-    private void success(long sessionId, int serverSessionStreamId) {
-      promise.onNext(new ConnectAckResponse(sessionId, serverSessionStreamId));
-      promise.onComplete();
-    }
-
-    @Override
-    public void dispose() {
-      connectAckPromises.remove(connectRequestId);
-      promise.cancel();
-    }
-
-    @Override
-    public boolean isDisposed() {
-      return promise.isDisposed();
-    }
-  }
-
-  /**
-   * Tuple containing connection ack response from server which are: session id and server session
-   * stream id. DTO class.
-   */
-  private static class ConnectAckResponse {
-
-    private final long sessionId;
-    private final int serverSessionStreamId;
-
-    private ConnectAckResponse(long sessionId, int serverSessionStreamId) {
-      this.sessionId = sessionId;
-      this.serverSessionStreamId = serverSessionStreamId;
+  private void setupConnection(
+      int sessionId, DefaultAeronConnection connection, MonoProcessor<Void> disposeHook) {
+    // listen shutdown
+    disposeHook
+        .then(Mono.fromRunnable(connection::dispose).then(connection.onDispose()))
+        .doFinally(
+            s -> logger.debug("{}: client connection disposed", Integer.toHexString(sessionId)))
+        .subscribe(
+            null,
+            th -> {
+              // no-op
+            });
+
+    if (handler == null) {
+      logger.warn("{}: handler function is not specified", Integer.toHexString(sessionId));
+      return;
+    }
+
+    try {
+      if (!connection.isDisposed()) {
```

Comment:
is it possible on this stage? if no let`s merge `disposeHook.subscribe` with `handler.subscribe` into one